File: 992.py


class Solution:    
    def SlidingWindow(self, nums: List[int], k: int)-> int:
        left, right, ans = 0, 0, 0
        checkDic = {}
        while right < len(nums):
            checkDic[nums[right]] = checkDic.get(nums[right], 0) + 1
            while len(checkDic) > k:
                checkDic[nums[left]] -= 1
                if checkDic[nums[left]] == 0:
                    del checkDic[nums[left]]
                left += 1
            ans += right - left + 1
            right += 1
        return ans 

    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        return self.SlidingWindow(nums,k) - self.SlidingWindow(nums, k-1)

    

File: 992.cpp
class Solution {
public:
    int SlidingWindow(vector<int>& nums, int k) {
        int left = 0, right = 0, ans = 0;
        unordered_map<int,int> check;
        while (right < nums.size())
        {
            check[nums[right]]++;
            while (check.size() > k){
                check[nums[left]]--;
                if(check[nums[left]] == 0)
                check.erase(nums[left]);
                left++;
            }
            ans+= right - left +1;
            right++;
        }
        return ans;
    }


    int subarraysWithKDistinct(vector<int>& nums, int k) {
        return SlidingWindow(nums,k) - SlidingWindow(nums, k-1);
    }
};